drivers/gles2/rasterizer_canvas_gles2.cpp

```diff
@@ -541,818 +554,1348 @@ static const GLenum gl_primitive[] = {
 
-				Item::CommandRect *r = static_cast<Item::CommandRect *>(command);
+				// 10% faster calculating the max first
+				Vector2 pos_max = src_rect.position + src_rect.size;
+				Vector2 uvs[4] = {
+					src_rect.position,
+					Vector2(pos_max.x, src_rect.position.y),
+					pos_max,
+					Vector2(src_rect.position.x, pos_max.y),
+				};
+
+				if (r->flags & CANVAS_RECT_TRANSPOSE) {
+					SWAP(uvs[1], uvs[3]);
+				}
+
+				if (r->flags & CANVAS_RECT_FLIP_H) {
+					SWAP(uvs[0], uvs[1]);
+					SWAP(uvs[2], uvs[3]);
+				}
+				if (r->flags & CANVAS_RECT_FLIP_V) {
+					SWAP(uvs[0], uvs[3]);
+					SWAP(uvs[1], uvs[2]);
+				}
 
-				glDisableVertexAttribArray(VS::ARRAY_COLOR);
-				glVertexAttrib4fv(VS::ARRAY_COLOR, r->modulate.components);
+				bA->uv.set(uvs[0]);
+				bB->uv.set(uvs[1]);
+				bC->uv.set(uvs[2]);
+				bD->uv.set(uvs[3]);
 
-				bool can_tile = true;
-				if (r->texture.is_valid() && r->flags & CANVAS_RECT_TILE && !storage->config.support_npot_repeat_mipmap) {
-					// workaround for when setting tiling does not work due to hardware limitation
+				// increment quad count
+				quad_count++;
 
-					RasterizerStorageGLES2::Texture *texture = storage->texture_owner.getornull(r->texture);
+			} break;
+		}
+	}
 
-					if (texture) {
+	// gotos are cool, never use goto kids
+cleanup:;
 
-						texture = texture->get_ptr();
+	// some heuristic to decide whether to use colored verts.
+	// feel free to tweak this.
+	// this could use hysteresis, to prevent jumping between methods
+	// .. however probably not necessary
+	if ((color_changes * 4) >= (quad_count * 3)) {
+		bdata.use_colored_verts = true;
 
-						if (next_power_of_2(texture->alloc_width) != (unsigned int)texture->alloc_width && next_power_of_2(texture->alloc_height) != (unsigned int)texture->alloc_height) {
-							state.canvas_shader.set_conditional(CanvasShaderGLES2::USE_FORCE_REPEAT, true);
-							can_tile = false;
-						}
-					}
+		// small perf cost versus going straight to colored verts (maybe around 10%)
+		// however more straightforward
+		_batch_translate_to_colored();
+	} else {
+		bdata.use_colored_verts = false;
+	}
+
+	// important, we return how far we got through the commands.
+	// We may not yet have reached the end, the vertex buffer may be full,
+	// and need a draw / reset / fill / rinse repeat
+	return command_num;
+}
+
+// convert the stupidly high amount of batches (each with its own color)
+// to larger batches where the color is stored in the verts instead...
+// There is a trade off. Non colored verts are smaller so work faster, but
+// there comes a point where it is better to just use colored verts to avoid lots of
+// batches.
+void RasterizerCanvasGLES2::_batch_translate_to_colored() {
+	bdata.verts_colored.reset();
+	bdata.batches_temp.reset();
+
+	CRASH_COND(bdata.verts_colored.max_size() != bdata.verts.max_size());
```

Comment:
They actually should be debug only, I've wrapped them in #ifdef DEBUG_ENABLED. It is a check on program flow, if this assertion is violated the rest of the function won't work. There is no sensible way to recover. Good news is users can't cause this, only a core programmer changing the allocation sizes so they are unequal, and it would be best to catch this immediately.